Binary data transfer protocol over ipv8.

drew2a opened the pull request:
This PR makes it possible to transfer big binary data over ipv8.
Closes https://github.com/Tribler/tribler/issues/6003

The protocol based on TFTP with windowsize (RFC 7440).

Features:

 * timeout
 * retransmit
 * dynamic window size

The maximum data size that can be transferred through the protocol can be calculated as "block_size * 4294967295" where 4294967295 is the max segment number (4B unsigned int). But the protocol is slow by design, so try to send less rather than more.

### How to use:

```python
class MyCommunity(EVAProtocolMixin, Community):
    community_id = b'0' * 20

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.eva_init()

        self.eva_register_receive_callback(self.on_receive)
        self.eva_register_send_complete_callback(self.on_send_complete)
        self.eva_register_error_callback(self.on_error)

    def my_function(self, peer):
        self.eva_send_binary(peer, b'info1', b'data1', 42)
        self.eva_send_binary(peer, b'info2', b'data2', 42)
        self.eva_send_binary(peer, b'info3', b'data3', 42)

   def on_receive(self, peer, binary_info, binary_data, nonce):
        logger.info(f'Data has been received: {binary_info}')

   def on_send_complete(self, peer, binary_info, binary_data, nonce):
        logger.info(f'Transfer has been completed: {binary_info}')

   def on_error(self, peer, exception, nonce):
        logger.error(f'Error has been occurred: {exception}')
```


Due to ipv8 specifics, we can use only one socket port per one peer. Therefore, at one point in time, the protocol can only transmit one particular piece of data for one particular peer.

In case the `eva_send_binary` method is invoked multiply times for a single peer, the data transfer will be scheduled and performed when the current transfer session is finished.

qstokkink commented on src/tribler-core/tribler_core/modules/metadata_store/community/eva_protocol.py:
```diff
@@ -0,0 +1,502 @@
+            nonce=0,
+        )
+
+        # specify messages' ids
+        _assign_id_to_messages(start_message_id, [WriteRequest, Acknowledgement, Data, Error])
+
+        # add messages' handlers
+        self.add_message_handler(WriteRequest, self.on_eva_write_request)
+        self.add_message_handler(Acknowledgement, self.on_eva_acknowledgement)
+        self.add_message_handler(Data, self.on_eva_data)
+        self.add_message_handler(Error, self.on_eva_error)
```
Quick note: you don't need to change the `msg_id` of the class instances (which could screw with other community instances doing the same). You can specify the handlers by message number instead:

```suggestion
        # add messages' handlers
        self.add_message_handler(start_message_id, self.on_eva_write_request)
        self.add_message_handler(start_message_id + 1, self.on_eva_acknowledgement)
        self.add_message_handler(start_message_id + 2, self.on_eva_data)
        self.add_message_handler(start_message_id + 3, self.on_eva_error)
```

qstokkink replied:
My bad, forgot to mention that you're not allowed to use `self.ez_send()` anymore then.

```python
self.ez_send(peer, Acknowledgement(0, transfer.nonce))
```
⬇️ (with `self.ack_msg_id = start_message_id + 1`)
```python
self.endpoint.send(peer.address, self.ezr_pack(self.ack_msg_id, Acknowledgement(0, transfer.nonce)))
```

drew2a replied:
I see. 
I'll rewrite all the occurrences because probably it seems like a single correct way to implement a dynamic message-id assignation (in case this mixin can be used by multiple classes in the future).

qstokkink commented:
Some further high-level comments while this is still in WIP:

 - Using `asyncio.create_task` is dangerous. This (a) schedules a task outside of the IPv8 error handler and consequently (b) doesn't cancel tasks when the Community is unloaded (calling dead objects usually leads to weird stuff on shut down). You're encouraged to use `self.register_anonymous_task` instead to bind the task to the Community.

 - Binding to `self.community*` and `self.peer*` in the tests is unnecessary. For this the `self.overlay(*)` and `self.peer(*)` exist. On another note, if you use `my_peer` from a community you're using the `Peer` instance that has a private key instead of a public key, which can lead to issues (`self.peer(*)` provides the public key view of a peer).

 - TestBase instantiates a new Community instance per test run. There is no need to remember the overwritten instance methods and restore them in a `finally` clause (for example the callbacks in `decode_map`). Also there doesn't seem to be a need to call `clear()` for the same reason.

 - Storing states in `dict`s almost always leads to memory leaks. You're encouraged to use [RequestCache](https://py-ipv8.readthedocs.io/en/latest/basics/requestcache_tutorial/#the-easier-way)s to keep track of states. These automatically time out and IPv8 can retrieve them in automatically in your message flow using the decorator. This also downsizes your code and makes it less complex.

 - I'd encourage the use of `asyncio` `Future`s for your calls. This allows external code to, for example, `await eva_send_binary()`. This will also downsize your tests. The `RequestCache` `Caches` have the `register_future()` to make sure `Future`s are also properly canceled when the overlay is destroyed.

 - (minor) `drain_loop` seems functionally equivalent to `self.deliver_messages(0.0)`.

drew2a commented:
@synctext yes, sure. Tests already there: https://github.com/Tribler/tribler/pull/6021/files#diff-0e1f354d589644baf35d810ec6117405b591e2428200678ac38fd8abab1b23a6R319-R377

drew2a commented:
@qstokkink thank you for the comments. 
I will work on them today.

>I'd encourage the use of asyncio Futures for your calls. This allows external code to, for example, await eva_send_binary(). This will also downsize your tests. The RequestCache Caches have the register_future() to make sure Futures are also properly canceled when the overlay is destroyed.

My original version of the protocol's interface was with `async` functionality. But @ichorid strongly asked me to rewrite it with normal calls.

I'm more in favor of the `async` version, BTW.

drew2a commented:
@qstokkink thanks for the comments.

> Using asyncio.create_task is dangerous. This (a) schedules a task outside of the IPv8 error handler and consequently (b) doesn't cancel tasks when the Community is unloaded (calling dead objects usually leads to weird stuff on shut down). You're encouraged to use self.register_anonymous_task instead to bind the task to the Community.

Implemented.

> Binding to self.community* and self.peer* in the tests is unnecessary. For this the self.overlay(*) and self.peer(*) exist. On another note, if you use my_peer from a community you're using the Peer instance that has a private key instead of a public key, which can lead to issues (self.peer(*) provides the public key view of a peer).

Implemented

> TestBase instantiates a new Community instance per test run. There is no need to remember the overwritten instance methods and restore them in a finally clause (for example the callbacks in decode_map). Also there doesn't seem to be a need to call clear() for the same reason.

Implemented

> Storing states in dicts almost always leads to memory leaks. You're encouraged to use RequestCaches to keep track of states. These automatically time out and IPv8 can retrieve them automatically in your message flow using the decorator. This also downsizes your code and makes it less complex.

About this. At the moment I continue to use dicts because I don't see the real need to change them to RequestCaches.
But it is still an open question.

qstokkink commented:
Cool 👍 

> At the moment I continue to use dicts because I don't see the real need to change them to RequestCaches.

Well you don't strictly _need_ them. They are just an easier way to store states of message flows and deal with all of the possible unhappy flows that can cause data to stick around in your dicts.

Without RequestCaches you'll have to manually make sure your dicts are cleaned up if any response message does not arrive (or crashes mid-handler or whatever). Historically, we've not been very good at this. We've spent quite some time digging through memory dumps (which is not fun).

qstokkink reviewed (approved):
I already gave my high-level comments and apart from that I only have some minor comments left that you could consider for future work/PRs:

 - The order in which `_eva_register_message_handler` is called defines the message wire format. I'd add a comment here so a future developer doesn't order the handlers alphabetically.
 - The docstring examples don't follow RST styling and won't be rendered correctly.
 - 10~20 second unit tests are not ideal.
 - I'd prefer `logging.info()` over `print()` in the tests (which is the recommended style approach by most code style tools). This also gets rid of the `# noqa: T001`.

I don't see any game-breakers and in my opinion this PR is therefore good to merge.

drew2a commented:
@qstokkink thank you 🙏 

> The order in which _eva_register_message_handler is called defines the message wire format. I'd add a comment here so a future developer doesn't order the handlers alphabetically.

Implemented

> The docstring examples don't follow RST styling and won't be rendered correctly.

Correct. But there is no need to render them (am I wrong?). 

> 10~20 second unit tests are not ideal.

Correct, but all tests here are not unit tests (it is also correct for almost all descendants of TestBase). 
Anyway, the tests should last as little time as possible.
I decrease the default time for timeouts and retransmit. It should decrease the execution time of the tests.

> I'd prefer logging.info() over print() in the tests (which is the recommended style approach by most code style tools). This also gets rid of the # noqa: T001.

Implemented.

Once again, thank you for the feedback.

qstokkink commented:
Nice, I'll re-approve momentarily.

>> The docstring examples don't follow RST styling and won't be rendered correctly.
>
> Correct. But there is no need to render them (am I wrong?).

If you press Ctrl+Q in PyCharm you get the RST rendering. Also, we used to automatically generate the documentation from the classes for ReadTheDocs.. however: https://tribler.readthedocs.io/en/latest/py-modindex.html kinda broken apparently 😃

ichorid reviewed (approved):
After three iterations, almost perfect :+1:
